### PR TITLE
ddl: delete ranges when rolling back 'add index'

### PR DESCRIPTION
**Pull request by zimulala**
We need to delete ranges when the operation of 'add index' needs to be rolled back.

### CODE REVIEW

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -155,6 +155,12 @@ func (d *ddl) finishDDLJob(t *meta.Meta, job *model.Job) (err error) {
 		metrics.DDLWorkerHistogram.WithLabelValues(metrics.WorkerFinishDDLJob, metrics.RetLabel(err)).Observe(time.Since(startTime).Seconds())
 	}()
 	switch job.Type {
+	case model.ActionAddIndex:
+		if job.State != model.JobStateRollbackDone {
+			break
+		}
+		// ActionAddIndex needs to delete ranges when it needs to be rolled back.
+		fallthrough
```
Comment (coocood) [inline ddl/ddl_worker.go]: `fallthrough` is rarely used, it's hard to read.
How about extracting a function `shouldDeleteRange`?

File: ddl/index.go [hunk 2]
```diff
@@ -395,6 +395,7 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 			job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
 		}
 		job.Args = append(job.Args, indexInfo.ID)
+		log.Warnf("args %v, ts %v", job.Args, t.StartTS)
```
Comment (coocood) [inline ddl/index.go]: Remove the `warn` log.

**Comment by zimulala** [comment]
PTAL @coocood

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -147,23 +147,36 @@ func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, meetErr bool) error {
...
-		} else {
-			err = errInvalidJobVersion.GenByArgs(job.Version, currentVersion)
-		}
-		if err != nil {
-			return errors.Trace(err)
+	case model.ActionAddIndex:
+		if job.State != model.JobStateRollbackDone {
+			break
 		}
+		// ActionAddIndex needs to delete ranges when it needs to be rolled back.
```
Comment (shenli) [inline ddl/ddl_worker.go]: After rolling back an AddIndex operation, we need to use delete-range to delete the half-done index data.

**Comment by zimulala** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@winkyao @coocood @ciscoxll PTAL

**Pull request merged**